modules/plugin/jdbc/jdbc-postgis/src/test/java/org/geotools/data/postgis/PostgisJNDIDataSourceOnlineTest.java

```diff
@@ -52,6 +52,7 @@ public void testExtraParams() {
         standardParams.remove(PostgisNGDataStoreFactory.CREATE_DB_IF_MISSING.key);
         standardParams.remove(PostgisNGDataStoreFactory.CREATE_PARAMS.key);
         standardParams.remove(PostgisNGDataStoreFactory.SSL_MODE.key);
+        standardParams.remove(PostgisNGDataStoreFactory.REWRITE_BATCHED_INSERTS.key);
```

Comment:
So I am not seeing any test coverage of this new feature? Can you add a method that uses your code (we want to be sure that something is actually happening and that it produces a SQL Insert statement that works).

Reply:
I think this is all happening inside the driver, so you won't see any difference in the SQL generated by JDBCDataStore. Nevertheless, it would be useful to have one write test run with these settings, to prove they work. Given that prepared statements are already batching write requests, I'm guessing this may be  useful for the case when prepared statements are not enabled. 

@j-heaton can you make sure there is a subclass of PostgisFeatureStoreOnlineTest that forces the new flag on? Also, did you measure performance improvements with the new settings?
How does it compare with the case where prepared statements are enabled?

Reply:
you are correct that it happens inside the driver. i was able to verify the behavior by setting `log_statements = 'all'` on the postgres DB and checking logs.

i have added a subclass of PostgisFeatureStoreOnlineTest that sets the new flag.

i was able to measure the performance and saw significant gains in ingest speed. i wrote a simple test to prove this although i'm not sure if it belongs in a commit.
couple caveats for this:
- statements will NOT be rewritten if geotools is generating fids for features. therefore features need to already have a fid and the hints `PROVIDED_FID` and `USE_PROVIDED_FID` must be used.
- `BATCH_INSERT_SIZE` needs to be increased or else performance will be worse than the case where `REWRITE_BATCHED_INSERTS` is disabled

here is the test i used along with the results:
```
REWRITE_BATCHED_INSERTS=false BATCH_INSERT_SIZE=100
  avg: 1989.375 min: 1875 max: 2125
REWRITE_BATCHED_INSERTS=false BATCH_INSERT_SIZE=1000
  avg: 1367.375 min: 1343 max: 1398
REWRITE_BATCHED_INSERTS=true BATCH_INSERT_SIZE=100
  avg: 2449.0 min: 2375 max: 2620
REWRITE_BATCHED_INSERTS=true BATCH_INSERT_SIZE=1000
  avg: 895.25 min: 872 max: 916
```
  
```
package org.geotools.data.postgis.ps;

import java.io.IOException;
import java.sql.SQLException;
import java.util.*;

import org.geotools.api.data.DataStore;
import org.geotools.api.data.Transaction;
import org.geotools.api.feature.simple.SimpleFeature;
import org.geotools.api.feature.simple.SimpleFeatureType;
import org.geotools.data.DefaultTransaction;
import org.geotools.data.postgis.PostgisNGDataStoreFactory;
import org.geotools.data.simple.SimpleFeatureCollection;
import org.geotools.feature.DefaultFeatureCollection;
import org.geotools.feature.SchemaException;
import org.geotools.feature.simple.SimpleFeatureBuilder;
import org.geotools.feature.simple.SimpleFeatureTypeBuilder;
import org.geotools.jdbc.JDBCFeatureStore;
import org.geotools.referencing.crs.DefaultGeographicCRS;
import org.geotools.test.OnlineTestSupport;
import org.geotools.util.factory.Hints;
import org.junit.AfterClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.locationtech.jts.geom.Coordinate;
import org.locationtech.jts.geom.GeometryFactory;
import org.locationtech.jts.geom.Point;

@RunWith(Parameterized.class)
public class PostGISReWriteBatchedInsertsPerfOnlineTest extends OnlineTestSupport {
    private static final String TYPE_NAME = "PostgisBatchInsertTest";
    private static final int TOTAL_FEATURES = 100000;
    private static final int NUM_FEATURE_ATTRIBUTES = 0;
    private static final int ITERATIONS = 8;

    private static final List<SimpleFeature> preGeneratedFeatures = generateFeatures();
    private static final List<Stats> results = new ArrayList<>();

    private final boolean reWriteBatchedInserts;
    private final int batchSize;
    @Parameterized.Parameters(name = "rewrite batches={0}, batch size={1}")
    public static Collection<Object[]> data() {
        return Arrays.asList(new Object[][] {
                { false, 100 },
                { false, 1000 },
                { true, 100 },
                { true, 1000 }
        });
    }

    public PostGISReWriteBatchedInsertsPerfOnlineTest(boolean reWriteBatchedInserts, int batchSize) {
        this.reWriteBatchedInserts = reWriteBatchedInserts;
        this.batchSize = batchSize;
    }

    @Override
    protected String getFixtureId() {
        return "postgis";
    }

    @Test
    public void testReWriteBatchedInserts() throws IOException, SQLException, SchemaException {
        System.out.printf("running test with REWRITE_BATCHED_INSERTS=%s, BATCH_INSERT_SIZE=%s\n", reWriteBatchedInserts, batchSize);
        System.out.println("warm up");
        Stats stats = new Stats(reWriteBatchedInserts, batchSize);
        runBatchInsertTest(reWriteBatchedInserts, batchSize, stats);
        stats.reset();
        System.out.println("begin recording iterations");
        for (int i=0; i<ITERATIONS; i++) {
            runBatchInsertTest(reWriteBatchedInserts, batchSize, stats);
        }
        System.out.printf("iteration stats:\n  avg: %sms\n  min: %sms\n  max: %sms\n\n", stats.getAverage(), stats.getMin(), stats.getMax());
        results.add(stats);
    }

    @AfterClass
    public static void summarizeResults() {
        System.out.println("all tests complete. results:");
        for (Stats result : results) {
            System.out.printf("REWRITE_BATCHED_INSERTS=%s BATCH_INSERT_SIZE=%s\n  avg: %s min: %s max: %s\n", result.isReWriteBatchedInserts(), result.getBatchSize(), result.getAverage(), result.getMin(), result.getMax());
        }
    }

    private static List<SimpleFeature> generateFeatures() {
        List<SimpleFeature> features = new ArrayList<>();
        SimpleFeatureType featureType = buildFeatureType();
        SimpleFeatureBuilder featureBuilder = new SimpleFeatureBuilder(featureType);
        GeometryFactory geometryFactory = new GeometryFactory();
        for (int i = 0; i < PostGISReWriteBatchedInsertsPerfOnlineTest.TOTAL_FEATURES; i++) {
            featureBuilder.set("geom", geometryFactory.createPoint(new Coordinate(i, i)));
            for (int j = 0; j < NUM_FEATURE_ATTRIBUTES; j++) {
                featureBuilder.set("field_" + j, String.valueOf(j));
            }
            featureBuilder.featureUserData(Hints.PROVIDED_FID, "fid");
            featureBuilder.featureUserData(Hints.USE_PROVIDED_FID, Boolean.TRUE);
            features.add(featureBuilder.buildFeature(String.valueOf(i)));
        }

        return features;
    }

    private void runBatchInsertTest(boolean reWriteBatchedInserts, int batchSize, Stats stats) throws IOException {
        DataStore store = createDataStore(reWriteBatchedInserts, batchSize);
        SimpleFeatureType featureType = buildFeatureType();
        try {
            store.removeSchema(TYPE_NAME);
        } catch (Exception ignored) { }
        store.createSchema(featureType);
        JDBCFeatureStore featureStore = (JDBCFeatureStore) store.getFeatureSource(TYPE_NAME);
        Transaction tx = new DefaultTransaction();
        DefaultFeatureCollection collection = new DefaultFeatureCollection();
        collection.addAll(preGeneratedFeatures);

        long startTime = System.currentTimeMillis();
        try {
            featureStore.setTransaction(tx);
            featureStore.addFeatures((SimpleFeatureCollection) collection);
            tx.commit();
        } catch (Exception e) {
            e.printStackTrace();
            tx.rollback();
            throw e;
        } finally {
            tx.close();
        }
        long endTime = System.currentTimeMillis();
        long totalTime = endTime - startTime;

        System.out.printf("iteration completed in %sms\n", totalTime);
        stats.update(totalTime);
        store.removeSchema(TYPE_NAME);
        store.dispose();
    }

    private DataStore createDataStore(boolean reWriteBatchedInserts, int batchSize) throws IOException {
        PostgisNGDataStoreFactory factory = new PostgisNGDataStoreFactory();
        Properties db = getFixture();
        Map<String, Object> params = new HashMap<>();

        params.put(PostgisNGDataStoreFactory.HOST.key, db.getProperty(PostgisNGDataStoreFactory.HOST.key));
        params.put(PostgisNGDataStoreFactory.DATABASE.key, db.getProperty(PostgisNGDataStoreFactory.DATABASE.key));
        params.put(PostgisNGDataStoreFactory.PORT.key, db.getProperty(PostgisNGDataStoreFactory.PORT.key));
        params.put(PostgisNGDataStoreFactory.USER.key, db.getProperty(PostgisNGDataStoreFactory.USER.key));
        params.put(PostgisNGDataStoreFactory.PASSWD.key, db.getProperty(PostgisNGDataStoreFactory.PASSWD.key));
        params.put(PostgisNGDataStoreFactory.DBTYPE.key, "postgis");

        params.put(PostgisNGDataStoreFactory.PREPARED_STATEMENTS.key, Boolean.TRUE);

        params.put(PostgisNGDataStoreFactory.REWRITE_BATCHED_INSERTS.key, reWriteBatchedInserts);
        params.put(PostgisNGDataStoreFactory.BATCH_INSERT_SIZE.key, batchSize);

        return factory.createDataStore(params);
    }

    private static SimpleFeatureType buildFeatureType() {
        SimpleFeatureTypeBuilder typeBuilder = new SimpleFeatureTypeBuilder();
        typeBuilder.setName(TYPE_NAME);
        DefaultGeographicCRS crs = DefaultGeographicCRS.WGS84;
        typeBuilder.setCRS(crs);
        typeBuilder.add("geom", Point.class);
        for (int j = 0; j < NUM_FEATURE_ATTRIBUTES; j++) {
            typeBuilder.add("field_" + j, String.class);
        }
        return typeBuilder.buildFeatureType();
    }

    private static class Stats {
        private final boolean reWriteBatchedInserts;
        private final int batchSize;

        private float totalTime = 0;
        private long minTime = Long.MAX_VALUE;
        private long maxTime = Long.MIN_VALUE;
        private int count = 0;

        public Stats(boolean reWriteBatchedInserts, int batchSize) {
            this.reWriteBatchedInserts = reWriteBatchedInserts;
            this.batchSize = batchSize;
        }

        public void update(long time) {
            totalTime += time;
            minTime = Math.min(minTime, time);
            maxTime = Math.max(maxTime, time);
            count++;
        }

        public boolean isReWriteBatchedInserts() {
            return reWriteBatchedInserts;
        }

        public int getBatchSize() {
            return batchSize;
        }

        public float getAverage() {
            return count > 0 ? totalTime / count : 0;
        }

        public long getMin() {
            return minTime;
        }

        public long getMax() {
            return maxTime;
        }

        public void reset() {
            totalTime = 0;
            minTime = Long.MAX_VALUE;
            maxTime = Long.MIN_VALUE;
            count = 0;
        }
    }

}
```